README.md
# CDISC Library API Mirror

This script allows to use CDISC Library API by keeping the authentication information in a central location.

**WARNING** You must follow the CDISC Library API EULA, including the following:
* You shall not (and will not allow or assist any third party to) access the API in order to replicate or compete with CDISC Library or CDISC Materials.
* You are solely responsible for the user(s) accessing the API using the account provided

* Verify that the server at which you are running the application is available only in the intranet.*


# Installation
### Prerequisites

The following software is needed to compile the application:
* Git
* Node.js
* Yarn

### Configuration
Create file **clamirror.conf** in the project root folder (see clamirror.conf.example):
```
{
    "auth": {
        "username": "CDISC Library API username",
        "password": "CDISC Library API password"
    },
    "port": 4600,
    "cache": {
        "enabled": false,
        "includeFilter": [],
        "excludeFilter": ["health", "mdr/lastupdated", "mdr/products", ".*/root/.*", "mdr/search/.*"],
        "cacheFolder": "/path/to/cache/folder"
    }
    "cors": {
        "enabled": false,
        "origins": ["*"]
    }
}
```

The mirror support basic caching functionality, to use it, change ***cache.enabled*** value to true.
* cache.enabled - controls whether cache is used.
* cache.includeFilter - array of endpoint regexes which are cached. If none specified, all endpoints are cached, except for those specified in excludeFilter.
* cache.excludeFilter - array of endpoint regexes which are not cached. It is suggested to use values from the example above.
* cache.cacheFolder - path to the folder where cached values will be stored.
* cors.enalbed - controls whether cross-origin resource sharing is used.
* cors.origins - array of origins which are allowed. Value "*" allows all origins.

### Installation

Clone the repository:
```
git clone https://github.com/defineEditor/cla-mirror.git
```
Navigate to the downloaded folder and install all required dependencies:
```
yarn install
```
Run the script:
```
yarn start
```
Now you should be able to access the API via the port specified in the configuration file, e.g.:
```
your.domain.com:4600/api/mdr/products
```

You can use packages like [pm2](https://www.npmjs.com/package/pm2), which will help to run the process as a daemon. In this case instead of using *yarn start*, you can start it as:
```
pm2 start src/app.js
```

### Additions

You can access CSV and XLS formats by adding '&format=(csv|xls|json)' at the end of the address

```
your.domain.com:4600/api/mdr/adam/adam-occds-1-0&format=csv
```
Note that only a limited number of API endpoints have CSV or XLS formats. See [CDISC Library documentation](https://wiki.cdisc.org/display/LIBSUPRT/Documentation) for details.

## Authors

* [**Dmitry Kolosov**](https://www.linkedin.com/in/dmitry-kolosov-91751413/)
* [**Sergei Krivtcov**](https://www.linkedin.com/in/sergey-krivtsov-677419b4/)

## License

This project is licensed under the MIT License - see the [LICENSE](LICENSE) file for details


CHANGELOG.md
# Changes
See [Releases](https://github.com/defineEditor/cla-mirror/releases) for the list of changes.


src/cache.js
const fs = require('fs');
const Jszip =  require('jszip');
const crypto = require('crypto');
const { promisify } = require('util');

const readFile = promisify(fs.readFile);
const readdir = promisify(fs.readdir);
const mkdir = promisify(fs.mkdir);
const writeFile = promisify(fs.writeFile);
const stat = promisify(fs.stat);

class ClaCache {
    constructor ({cacheFolder, cachedIds = []} = {}) {
        this.cacheFolder = cacheFolder;
        this.cachedIds = cachedIds;
    }

    async init () {
        try {
            await mkdir(this.cacheFolder);
        } catch (err) {
            if (err.code === 'EEXIST') {
                // Folder exists, which is fine
            } else {
                let msg = 'Failed creating a cache folder ' + cacheFolder + '. Error: ' + err.message;
                return;
            }
        }
    
        // Get currently cached files
        let files;
        try {
            files = await readdir(this.cacheFolder);
        } catch (error) {
            return;
        }
    
        this.cachedIds = files.map(file => file.replace(/.gz$/,''));
    }

    async getRequestId (request) {
        let shortenedUrl = request.url
            .replace('/root/', '/r/')
            .replace('/cdash/', '/cd/')
            .replace('/cdashig/', '/cdi/')
            .replace('/sdtm/', '/s/')
            .replace('/sdtmig/', '/si/')
            .replace('/send/', '/se/')
            .replace('/sendig/', '/sei/')
            .replace('/adam/', '/a/')
            .replace('/root/', '/r/')
            .replace('/datasets/', '/d/')
            .replace('/domains/', '/dm/')
            .replace('/datastructures/', '/ds/')
            .replace('/classes/', '/c/')
            .replace('/variables/', '/v/')
            .replace('/fields/', '/f/')
            .replace('/varsets/', '/vs/')
            .replace('/packages/', '/p/')
            .replace('/codelists/', '/cl/')
            .replace('/terms/', '/t/')
            .replace('/scenarios/', '/s/')
            .replace(/.*?\/mdr\//, '')
        ;
        let requestOptions = JSON.toString({ ...request, url: undefined });
        let hashHex = crypto.createHash('sha1').update(requestOptions).digest('hex');
        if (hashHex === '0afd4c0de6a7b1a685edd9e8d152d66d5b4b7bd0') {
            // These are standard request options, no need to add a hash code
            return shortenedUrl;
        } else {
            return shortenedUrl + hashHex;
        }
    }

    async claMatch (request) {
        // Get an id
        let id = await this.getRequestId(request);

        // Search for the response in cache
        if (this.cachedIds.includes(id)) {
            let zippedData = await readFile(path.join(this.cacheFolder,id + '.gz'));
            let zip = new Jszip();
            await zip.loadAsync(zippedData);
            if (Object.keys(zip.files).includes('response.json')) {
                let result = await zip.file('response.json').async('string');
                return { statusCode: 200, body: result };
            }
        }
    }

    async claPut (request, response) {
        // Get an id
        let id = await this.getRequestId(request);
        // Minify the response
        let data = JSON.parse(response.body);
        // Compress the data
        let zip = new Jszip();
        zip.file('response.json', JSON.stringify(data));
        let zippedData = await zip.generateAsync({
            type: 'blob',
            compression: 'DEFLATE',
            compressionOptions: {
                level: 7
            }
        });
        await writeFile(path.join(this.cacheFolder,id + '.gz'), zippedData);
    }
}

module.exports = ClaCache;

src/claCache.js
const fs = require('fs');
const Jszip = require('jszip');
const path = require('path');
const { promisify } = require('util');

const readFile = promisify(fs.readFile);
const readdir = promisify(fs.readdir);
const mkdir = promisify(fs.mkdir);
const writeFile = promisify(fs.writeFile);

class ClaCache {
    constructor ({ cacheFolder, cachedIds = [], excludeFilter = [], includeFilter = [] } = {}) {
        this.cacheFolder = cacheFolder;
        this.cachedIds = cachedIds;
        this.excludeFilter = excludeFilter;
        this.includeFilter = includeFilter;
        // Bind methods
        this.init = this.init.bind(this);
        this.getRequestId = this.getRequestId.bind(this);
        this.filterRequest = this.filterRequest.bind(this);
        this.claMatch = this.claMatch.bind(this);
        this.claPut = this.claPut.bind(this);
    }

    async init () {
        try {
            await mkdir(this.cacheFolder);
        } catch (err) {
            if (err.code === 'EEXIST') {
                // Folder exists, which is fine
            } else {
                return;
            }
        }

        // Get currently cached files
        let files;
        try {
            files = await readdir(this.cacheFolder);
        } catch (error) {
            return;
        }

        this.cachedIds = files.map(file => file.replace(/.gz$/, ''));
    }

    getRequestId (request) {
        let shortenedUrl = request.url
            .replace(/^.*?api\//, '')
            .replace('/root/', '/r/')
            .replace('/cdash/', '/cd/')
            .replace('/cdashig/', '/cdi/')
            .replace('/sdtm/', '/s/')
            .replace('/sdtmig/', '/si/')
            .replace('/send/', '/se/')
            .replace('/sendig/', '/sei/')
            .replace('/adam/', '/a/')
            .replace('/root/', '/r/')
            .replace('/datasets/', '/d/')
            .replace('/domains/', '/dm/')
            .replace('/datastructures/', '/ds/')
            .replace('/classes/', '/c/')
            .replace('/variables/', '/v/')
            .replace('/fields/', '/f/')
            .replace('/varsets/', '/vs/')
            .replace('/packages/', '/p/')
            .replace('/codelists/', '/cl/')
            .replace('/terms/', '/t/')
            .replace('/scenarios/', '/s/')
            .replace(/.*?\/mdr\//, '')
            .replace(/\//g, '.')
        ;
        if (request && request.headers) {
            if (request.headers.Accept === 'application/json') {
                return shortenedUrl;
            } if (request.headers.Accept === 'application/text/csv') {
                return shortenedUrl + '.csv';
            } if (request.headers.Accept === 'application/vnd.ms-excel') {
                return shortenedUrl + '.excel';
            } else {
                return;
            }
        } else {
            return;
        }
    }

    filterRequest (request) {
        if (this.includeFilter.length > 0 || this.excludeFilter.length > 0) {
            let endpoint = request.url.replace(/^.*?api\//, '');
            let excludeMatched = this.excludeFilter.some(regex => (RegExp('^' + regex + '$').test(endpoint)));
            if (excludeMatched) {
                return true;
            }
            if (this.includeFilter.length > 0) {
                let includeMatched = this.includeFilter.some(regex => (RegExp('^' + regex + '$').test(endpoint)));
                if (!includeMatched) {
                    return true;
                }
            }
        }
        return false;
    }

    async claMatch (request) {
        // Do not match requests which are filtered
        if (this.filterRequest(request)) {
            return;
        }
        // Get an id
        let id = this.getRequestId(request);

        // Do not cache non-standard Accept header
        if (id === undefined) {
            return;
        }

        // Search for the response in cache
        if (this.cachedIds.includes(id)) {
            let zippedData = await readFile(path.join(this.cacheFolder, id + '.gz'));
            let zip = new Jszip();
            await zip.loadAsync(zippedData);
            if (Object.keys(zip.files).includes('response.json')) {
                let result = await zip.file('response.json').async('string');
                return JSON.parse(result);
            }
        }
    }

    async claPut (request, response) {
        // Do not store requests which are filtered
        if (this.filterRequest(request)) {
            return;
        }
        // Get an id
        let id = await this.getRequestId(request);
        // Minify the response
        let data = { headers: response.headers };
        if (data.headers['content-type'].startsWith('application/json')) {
            data.body = JSON.stringify(JSON.parse(response.body));
        } else {
            data.body = response.body;
        }
        // Compress the data
        let zip = new Jszip();
        zip.file('response.json', JSON.stringify(data));
        let zippedData = await zip.generateAsync({
            type: 'nodebuffer',
            compression: 'DEFLATE',
            compressionOptions: {
                level: 7
            }
        });
        await writeFile(path.join(this.cacheFolder, id + '.gz'), zippedData);
        this.cachedIds.push(id);
    }
}

module.exports = ClaCache;


src/wrapperRequest.js
const queryString = require('query-string');

const wrapperRequest = async (params, cdiscLibrary) => {
    let paramsParsed = queryString.parse(params);
    let { format, traffic, productList, product, productDetails, itemGroup, listItemGroups } = paramsParsed;

    // JSON is used by default when the format is not specified
    if (!format) {
        format = 'json';
    }

    if (traffic !== undefined) {
        return cdiscLibrary.getTrafficStats();
    }

    if (productList !== undefined) {
        return cdiscLibrary.getProductList(format);
    }

    if (productDetails !== undefined) {
        return cdiscLibrary.getProductDetails({ type: 'long', format });
    }

    if (product) {
        let prodId = await cdiscLibrary.getProductIdByAlias(product);
        if (!prodId) {
            return `No such product (${product}). See /?productList or /?productDetails for the full list of products.`;
        }
        let prd = await cdiscLibrary.getFullProduct(product, true);

        if (listItemGroups !== undefined) {
            return prd.getItemGroups({ type: 'short', format });
        }

        if (itemGroup) {
            let dataset = await prd.getItemGroup(itemGroup);
            if (dataset !== undefined) {
                return dataset.getFormattedItems(format);
            } else {
                return `No such itemGroup (${itemGroup}). See /?product=XXX&listItemGroups to get the full list of itemGroups.`;
            }
        }
    }
};

module.exports = wrapperRequest;


src/app.js
const express = require('express');
const fs = require('fs');
const os = require('os');
const path = require('path');
const compression = require('compression');
const { CdiscLibrary } = require('cla-wrapper');
const ClaCache = require('./claCache.js');
const wrapperRequest = require('./wrapperRequest.js');
const cors = require('cors');

var configData = fs.readFileSync(path.join(__dirname, '..', 'clamirror.conf'), 'utf8');
var config = JSON.parse(configData);

var app = express();

// Initialize a cache object
const cacheEnabled = config.cache && config.cache.enabled;
let claCache = {};
if (cacheEnabled) {
    claCache = new ClaCache({
        cacheFolder: config.cache.cacheFolder,
        excludeFilter: config.cache.excludeFilter,
        includeFilter: config.cache.includeFilter
    });
    claCache.init();
}

// Setup CORS
const corsEnabled = config.cors && config.cors.enabled;
if (corsEnabled) {
    const origins = config.cors.origins;
    let corsOptions;
    if (!(Array.isArray (origins) && origins.length === 1 && origins[0] === '*')) {
        corsOptions = {
            origin: function (origin, callback) {
                if (origins.indexOf(origin) !== -1) {
                    callback(null, true)
                } else {
                    callback(new Error('Not allowed by CORS'))
                }
            }
        }
    }
    app.use(cors(corsOptions));
}

// Adding compression
app.use(compression({ filter: (req, res) => {
    if (req.headers['x-no-compression']) {
        return false;
    } else if (typeof req.headers['accept-encoding'] === 'string') {
        if (req.headers['accept-encoding'].split(',').map(item => item.trim()).includes('gzip')) {
            return compression.filter(req, res);
        } else {
            return false;
        }
    } else {
        return false;
    }
}}));

// Load CLA Wrapper
const cl = new CdiscLibrary({
    username: config.auth.username,
    password: config.auth.password,
    cache: cacheEnabled ? { match: (request) => (claCache.claMatch(request)), put: claCache.claPut } : undefined,
});

app.use('/', async (req, res) => {
    try {
        if (req.url.startsWith('/api/')) {
            let endpoint = req.url.replace(/^\/api/, '');
            let headers = {};
            if (/application\/(json|vnd\.ms-excel)|text\/csv/i.test(req.headers.accept)) {
                headers.Accept = req.headers.accept;
            } else {
                // If format parameter is provided, use it to specify the response format
                if (/&format=csv/i.test(endpoint)) {
                    endpoint = endpoint.replace(/&format=csv/i, '');
                    headers = { 'Accept': 'text/csv' };
                } else if (/&format=xls/i.test(endpoint)) {
                    endpoint = endpoint.replace(/&format=xls/i, '');
                    headers = { 'Accept': 'application/vnd.ms-excel' };
                } else if (/&format=json/i.test(endpoint)) {
                    endpoint = endpoint.replace(/&format=json/i, '');
                }
            }
            let response = await cl.coreObject.apiRequest(
                endpoint,
                {
                    headers,
                    returnRaw: true,
                }
            );

            // If the request failed (undefined - value is taken from cache)
            if (response.statusCode !== 200 && response.statusCode !== undefined) {
                res.sendStatus(response.statusCode);
                return;
            }

            if (response.headers['content-type'] === 'application/vnd.ms-excel') {
                let fileName = endpoint.replace(/.*\/(.+\/.+)/, '$1').replace(/\W/g, '.') + '.xls';
                res.setHeader('Content-Disposition', 'attachment; filename=' + fileName);
                res.setHeader('Content-Type', response.headers['content-type']);
                res.setHeader('Transfer-Encoding', response.headers['transfer-encoding']);
                res.send(Buffer.from(response.body, 'binary'));
            } else {
                res.setHeader('Content-Type', response.headers['content-type']);
                res.send(response.body);
            }
        } else if (req.url.startsWith('/?')) {
            let response = await wrapperRequest(req.url.replace(/^\//, ''), cl);
            res.send(response.body);
        }
    } catch (error) {
        console.log(`Error when running a request for ${req.url}: ${error.message}`);
    }
});

var port = config.port || 4600;
app.listen(port, function () {
    console.log(`CDISC Library API Mirror is listening on port ${port}`);
});
